contracts/enclaves/AttestationAutherUpgradeable.sol

```diff
@@ -0,0 +1,156 @@
+
+    function _whitelistEnclaveImage(EnclaveImage memory image) internal returns(bytes32) {
+        require(
+            image.PCR0.length == 48 &&
+            image.PCR1.length == 48 &&
+            image.PCR2.length == 48,
+            "AA:WI-PCR values must be 48 bytes"
+        );
+
+        bytes32 imageId = keccak256(abi.encodePacked(image.PCR0, image.PCR1, image.PCR2));
+        require(whitelistedImages[imageId].PCR0.length == 0, "AA:WI-image already whitelisted");
+        whitelistedImages[imageId] = EnclaveImage(image.PCR0, image.PCR1, image.PCR2);
+        emit EnclaveImageWhitelisted(imageId, image.PCR0, image.PCR1, image.PCR2);
+        return imageId;
+    }
+
+    function _revokeEnclaveImage(bytes32 imageId) internal {
+        require(whitelistedImages[imageId].PCR0.length != 0, "AA:RI-Image not whitelisted");
+
+        delete whitelistedImages[imageId];
+        emit EnclaveImageRevoked(imageId);
+    }
+
+    function _whitelistEnclaveKey(bytes memory enclavePubKey, bytes32 imageId) internal {
+        require(whitelistedImages[imageId].PCR0.length != 0, "AA:WK-Image not whitelisted");
+        address enclaveKey = _pubKeyToAddress(enclavePubKey);
+        require(verifiedKeys[enclaveKey] == bytes32(0), "AA:WK-Enclave key already verified");
+
+        verifiedKeys[enclaveKey] = imageId;
+        emit EnclaveKeyWhitelisted(enclavePubKey, imageId);
+    }
+
+    function _revokeEnclaveKey(bytes memory enclavePubKey) internal {
+        address enclaveKey = _pubKeyToAddress(enclavePubKey);
+        require(verifiedKeys[enclaveKey] != bytes32(0), "AA:RK-Enclave key not verified");
+
+        delete verifiedKeys[enclaveKey];
+        emit EnclaveKeyRevoked(enclavePubKey);
+    }
+
+    // add enclave key of a whitelisted image to the list of verified enclave keys
+    function _verifyKey(
+        bytes memory signature,
+        bytes memory enclavePubKey,
+        bytes32 imageId,
+        uint256 enclaveCPUs,
+        uint256 enclaveMemory,
+        uint256 timestampInMilliseconds
+    ) internal {
+        require(
+            whitelistedImages[imageId].PCR0.length != 0,
+            "AA:VK-Enclave image to verify not whitelisted"
+        );
+        address enclaveKey = _pubKeyToAddress(enclavePubKey);
+        require(
+            verifiedKeys[enclaveKey] == bytes32(0),
+            "AA:VK-Enclave key already verified"
+        );
+        require(timestampInMilliseconds / 1000 > block.timestamp - ATTESTATION_MAX_AGE , "AA:VK-Attestation too old");
+
+        EnclaveImage memory image = whitelistedImages[imageId];
+        ATTESTATION_VERIFIER.verify(
+            signature,
+            enclavePubKey,
+            image.PCR0,
+            image.PCR1,
+            image.PCR2,
+            enclaveCPUs,
+            enclaveMemory,
+            timestampInMilliseconds
+        );
+
+        verifiedKeys[enclaveKey] = imageId;
+        emit EnclaveKeyVerified(enclavePubKey, imageId);
+    }
+
+    function verifyKey(
```

Comment:
Should this function exist as default, may be dev can choose to expose if required, probably with some restrictions.

Reply:
It's basically mandatory in any inheritor, no other way to really add keys otherwise (modulo the whitelist).

It's fine without any restrictions no? If it's really needed, can be overridden anyway to change behaviour.

Reply:
hmm... right. Makes sense.